### ISSUE TITLE
Regina rexx is never calling Hercules's Exit Handler!!

### ISSUE THREAD

**Issue by Fish-Git** [SDL-Hercules-390/hyperion#141]
Hercules is currently written in such a way that it does the _exact same thing_ in the _exact same way_ -- _regardless_ of which Rexx is being used. And _yet_, for some unknown reason, Regina never calls Hercules's "Exit Handler" whereas OORexx always does.

For the enlightened (fellow Hercules developers and developer wannabes), tests 3, 4 and 9 of the 3211 printer runtest are failing because Regina is, for reasons unknown, failing to honor our request for it to call our "Exit Handler".

It is Hercules's Exit Handler that intercepts rexx script "say" statements so that what the script says can be properly intercepted (handled/processed) by Hercules so that what is being said can be captured into the DIAG8 instruction's response buffer (which 3211 printer tests 3, 4 and 9 rely on)).

Because the Exit Handler is never being called by Regina, the 3211 printer test's DIAG8 response buffers are never filled in, thereby causing tests 3, 4 and 9 to fail.

I haven't confirmed it yet, but I strongly suspect it might have to do with Hercules's weird redirecting of 'stdout' and 'stderr' by the logger facility at startup/initialization.

Why that should matter to Regina I don't know, but apparently it does. OORexx works fine. Regina doesn't, and the stdout/stderr redirection that Hercules is doing seems to be the only remaining possibility!

Regina's documentation does seem to have special options for stdout/stderr and seems to particularly concerned about them, given how much attention they give to it. But as to _why_ it's so concerned about them I'll never know.

What is especially exasperating is, I wrote a simple test program (see further below) that calls RexxStart just like Hercules does, specifying the use of an Exit Handler (just like Hercules does), and believe it or not it works flawlessly for _both_ Regina _and_ OORexx!! (wtf?!)

So why doesn't it work when the _same exact thing_ is done from within Hercules?!

It's obvious something that Hercules is doing that is causing Regina to fail when called from within Hercules, but I haven't yet figured it out. As I said earlier, we're doing the _exact same thing_ for OORexx as we are for Regina, and yet OORexx works just fine with Hercules whereas Regina doesn't!

And the only reason I can think of is because of the weird stdout/stderr redirection (dup/dup2) we're doing at startup. It's the only thing left that I haven't tried yet!

Here's the simple test program I wrote:

* [RexxExit_Linux.zip](https://github.com/SDL-Hercules-390/hyperion/files/2338609/RexxExit_Linux.zip)&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;(for Linux only, obviously)
* [RexxExit_Windows.zip](https://github.com/SDL-Hercules-390/hyperion/files/2338610/RexxExit_Windows.zip)&nbsp;&nbsp;&nbsp;(same thing but for Windows)

The two versions (Linux vs. Windows) are essentially identical. The only difference is the #include headers and the loading of the libraries (LoadLibrary/GetProcAddress on Windows and dlopen/dlsym on Linux). Other than that they are identical.

I'm going to try modifying it to do the same weird redirection that Hercules is doing to see if that makes any difference to Regina. The only reason I haven't done that yet is because doing so is not as simple and straightforward as it may at first seem.

_**I need some help with this one, guys!**_&nbsp;&nbsp;&nbsp;`:((`

_(Please?)_

**Comment by Fish-Git** [comment]
Here is an _even better_ test that you can try for yourself that _**PROVES beyond a shadow of a doubt**_ that _**Regina Rexx contains serious bugs!**_ (or else Hercules does!)

* [regina-bug.zip](https://github.com/SDL-Hercules-390/hyperion/files/2342289/regina-bug.zip)

&nbsp;
Here is the text from the "README" file that is included in the above .zip file:
```
------------------------------------------------------------------------

                        REGINA REXX BUG PROOF

1. Apply the patch to current SDL Hyperion.

2. Rebuild Hercules.


The patch introduces some new flags in sysblk to help illustrate the bug.

The 'script' and 'exec' commands as well as the DIAGNOSE 0x008 instruction
all set the flags at the beginning of their logic, and then reset them
afterwards.

Each have a second flag which is initialized to FALSE. The Rexx 'ExitHandler'
function -- which we ALWAYS request when we ask Rexx to execute a script for
us -- sets the flag to TRUE.

When the 'script' or 'exec' command or the DIAG8 instruction completes,
they then each report on the state of this flag. If it's TRUE, then the
rexx 'ExitHandler' function was obviously called. If it's still FALSE,
then it obviously was not.


3. Copy the "d8rexx.rexx" and "d8rexx.txt" files to the ROOT (base) of the
   Hercules source code directory.  (NOT to the 'tests' subdirectory!)

4. Start Hercules and enter the following commands, one after the other:

      shcmdopt enable diag8
      diag8cmd enable noecho

      exec d8rexx.rexx
      script d8rexx.rexx
      script d8rexx.txt

5. Observe the results!


On systems with OORexx installed, they each work perfectly: each command
calls our rexx 'ExitHandler' function as we requested.

On systems with Regina installed however, only the first command -- the
'exec' command entered from the Herc command line -- works as expected.
Each of the other commands however -- the 'script' command and the DIAG8
test script -- FAIL to call our 'ExitHandler' function!

WHY?!  It's driving me nuts!

FURTHERMORE, on Windows, the text that the d8rexx.rexx script "say"s is
written to the Hercules panel's command-line! (input area)  Refer to the
"win_regina_bug.jpg" screen grab.

This only happens on Windows however (and only with Regina of course).
It doesn't seem to happen on Linux or with OORexx.

The four log files are the tests I've done on my Windows 7 system and on
my CentOS 6.10 system, each with ONLY OORExx installed and then again with
ONLY OORexx installed.  As you can see the tests work fine when OORexx is
installed and FAIL when Regina is installed.

Thanks for any insight/help you can provide into this mystery.

Fish
Sept. 1, 2018

------------------------------------------------------------------------
```

&nbsp;
Please try it for yourself, and then let me know what your thoughts are.

Should we drop Regina Rexx support altogether?

Or is there a unresolved bug somewhere in Hercules?

HELP!

Any insight any of you can provide would be greatly appreciated.

**Comment by mhoes** [comment]
Hi,

Not that this actually helps you, but ... You have expressed your suspicions that Hyperion's 'weird' (sorry, not sure what qualifies as that exactly) stdout/stderr redirection (dup/dup2) might be the cause of this issue.

So, I downloaded and compiled your 'rexxexit.c', and hacked up a small shell script that attempts to do something like that :

```
#!/bin/sh

# save stdout and stderr
exec 3>&1 4>&2

# redirect all output to a log file
exec > ./logfile.txt 2>&1

# execute rexxexit
./rexxexit rexxexit.rexx

# restore stdout and stderr
exec 1>&3 2>&4

# display logfile
cat ./logfile.txt
```

In an attempt to see if basic duplication and redirection of file descriptors reproduced the issue. (Granted, it's in the shell and not in the C code).

Which, according to 'strace' (the system call tracer) calls 'dup2()' (that would be the 'exec' statements in the shell script), and produces this output (which I guess is the correct output) :

```
Calling RunRexxScript( "rexxexit.rexx" )...

SIOExit: ENTRY
RXSIO == ExitNumber
RXSIOSAY: the script said: "REXX-Regina_3.9.1(MT) script "rexxexit.rexx" is running..."

SIOExit: ENTRY
RXSIO == ExitNumber
RXSIOSAY: the script said: "324"

Back from RunRexxScript()!
```

Which makes me wonder if the expressed suspicions are correct. Of course, it is very likely that this shell script does something entirely different than what the Hyperion code is doing, in which case this comment helps you even less.

**Comment by Fish-Git** [comment]
> What is especially exasperating is, I wrote a simple test program (see further below) that calls RexxStart just like Hercules does, specifying the use of an Exit Handler (just like Hercules does), and believe it or not it works flawlessly for _both_ Regina _and_ OORexx!! (wtf?!)
> 
> So why doesn't it work when the _same exact thing_ is done from within Hercules?!
> 
> It's obvious something that Hercules is doing that is causing Regina to fail when called from within Hercules, but I haven't yet figured it out. As I said earlier, we're doing the _exact same thing_ for OORexx as we are for Regina, and yet OORexx works just fine with Hercules whereas Regina doesn't!
> 
> And the only reason I can think of is because of the weird stdout/stderr redirection (dup/dup2) we're doing at startup. It's the only thing left that I haven't tried yet!

I've finally managed to add Hercules's redirection logic to my `rexxexit.c` test program and sure enough, now Regina is _failing to call the Exit Handler!_ So it's definitely Hercules's logger_init redirection logic that is confusing poor Regina. OORexx works flawlessly, with or without redirection, but Regina unfortunately doesn't.

I need to clean up my code a bit and try it on Linux as well (all my testing so far has been done on Windows) before posting it for others to download and try.

**Comment by Fish-Git** [comment]
> Of course, it is very likely that this shell script does something entirely different than what the Hyperion code is doing, ...

I can state with _100% certainty_ that your script is absolutely _**not**_ doing redirection the same way that Hercules is doing it.

&nbsp;
> ... in which case this comment helps you even less.

Bingo.&nbsp;`:(`

&nbsp;
_(I appreciate the effort though!)_

**Comment by Fish-Git** [comment]
I'm tired of farqing with this. Regina Rexx sucks. A workaround has already been committed so I see no need to waste anymore time with this issue. I'm closing it.

**Comment by mhoes** [comment]
> I've finally managed to add Hercules's redirection logic to my rexxexit.c test program and sure enough, now Regina is failing to call the Exit Handler! So it's definitely Hercules's logger_init redirection logic that is confusing poor Regina. OORexx works flawlessly, with or without redirection, but Regina unfortunately doesn't.
> 
> I need to clean up my code a bit and try it on Linux as well (all my testing so far has been done on Windows) before posting it for others to download and try.

Just wondering, did you ever get around to making that work on Linux as well ? (and if so, would you be willing to post it here ?)

**Comment by Fish-Git** [comment]
I still have the program (I think). Let me try to clean it up a bit and I'll post it here. Give me a day or so.

**Comment by Fish-Git** [comment]
> Let me try to clean it up a bit and I'll post it here. Give me a day or so.

I hope you know 'C', because I've changed my mind.

This whole issue just completely disgusts me and I no longer have any desire to work on it for even a few minutes. So I'm just going to give you the crap I've got and let _YOU_ (or someone else) screw around with it.

I found the programs, but the Windows version that contains the Hercules redirection logic (which requires threading logic too in order to do the redirection like Hercules does) also contains a bunch of multi-threading crap that I was farqing around with at the the time and is in a real mess. A BIG mess. A true state of disarray, and I don't feel like cleaning it up.

The Linux source is much cleaner and simpler because it was a much earlier version of my test program and doesn't have any redirection logic nor multi-threading logic in it:

Here they are:

* [reginabug.zip](https://github.com/SDL-Hercules-390/hyperion/files/2683246/reginabug.zip)

Good luck!
